Properly reset/adopt taskinstances and fix find_zombies

ephraimbuddy opened the pull request:
Currently, we fail long running SchedulerJob instances(LocalTaskJob) with no recent heartbeat
in reset_and_adopt_orphaned_tasks method of the Scheduler. Sometimes, this leads to the _find_zombies
method of the DAG file process manager running TaskCallbacks. This is how it happens:

When the scheduler dies and we reset task instances or adopt, we do not reset the ti.job_id and the
new SchedulerJob will create instances(LocalTaskJob) to run the new reset tasks. When the old job
which the task instances still has the job_id, has not heartbeated for long, the adopt_or_reset_orphaned
tasks will kill those jobs and subsequently or the _find_zombies method will find the tasks with the old jobID
and send a TaskCallback which will now fail the tasks being run by new schedulerjob instances

This was fixed by first, setting  ti.job_id to None
This helps to avoid clashing with the _find_zombies method which is now made to only act on jobs that are not
running and are also not heartbeating but the task is running.

In a situation where the _find_zombies method and adopt_or_reset_orphaned_task method run at the same time,
there would be conflict but now there won't be conflicts because _find_zombies only act on failed jobs with
running tasks and adopt_or_reset_orphaned_tasks kill jobs

Related: https://github.com/apache/airflow/issues/16023


---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

jedcunningham commented on airflow/jobs/scheduler_job.py:
```diff
@@ -1131,6 +1134,44 @@ def _emit_pool_metrics(self, session: Session = None) -> None:
+    @provide_session
+    def on_kill(self, session=None) -> None:
+        self.log.error("SchedulerJob failed. Job ID: %s", self.id)
+        self.state = State.FAILED
```
Don't we get here by the state being in a terminating state already? Is there another path?

jedcunningham commented on airflow/jobs/scheduler_job.py:
```diff
@@ -1131,6 +1134,44 @@ def _emit_pool_metrics(self, session: Session = None) -> None:
+    def on_kill(self, session=None) -> None:
+        self.log.error("SchedulerJob failed. Job ID: %s", self.id)
+        self.state = State.FAILED
+        self.end_date = timezone.utcnow()
```
Isn't this done in: https://github.com/apache/airflow/blob/0abbd2d918ad9027948fd8a33ebb42487e4aa000/airflow/jobs/base_job.py#L152

jedcunningham commented on airflow/jobs/scheduler_job.py:
```diff
@@ -1131,6 +1134,44 @@ def _emit_pool_metrics(self, session: Session = None) -> None:
+        if self.terminating:
+            return
+        if not self.is_alive():
+            self.on_kill()
```
```suggestion
            self.on_kill()
```

Should this call `.kill()` instead?

jedcunningham commented on airflow/jobs/scheduler_job.py:
```diff
@@ -1131,6 +1134,44 @@ def _emit_pool_metrics(self, session: Session = None) -> None:
+        self.log.error("SchedulerJob failed. Job ID: %s", self.id)
+        self.state = State.FAILED
+        self.end_date = timezone.utcnow()
+        self.terminating = True
```
Assuming we end up here because of:
https://github.com/apache/airflow/blob/0abbd2d918ad9027948fd8a33ebb42487e4aa000/airflow/jobs/base_job.py#L205-L206

Won't we not even make it to the next `heartbeat_callback`, meaning the `terminating` stuff you're adding won't be used? Am I missing something?

ephraimbuddy committed:
Properly reset/adopt taskinstances and fix find_zombies

Currently, we fail long running SchedulerJob instances(LocalTaskJob) with no recent heartbeat
in reset_and_adopt_orphaned_tasks method of the Scheduler. Sometimes, this leads to the _find_zombies
method of the DAG file process manager running TaskCallbacks. This is how it happens:

When the scheduler dies and we reset task instances or adopt, we do not reset the ti.job_id and the
new SchedulerJob will create instances(LocalTaskJob) to run the new reset tasks. When the old job
which the task instances still has the job_id, has not heartbeated for long, the adopt_or_reset_orphaned
tasks will kill those jobs and subsequently or the _find_zombies method will find the tasks with the old jobID
and send a TaskCallback which will now fail the tasks being run by new schedulerjob instances

This was fixed by first, setting  ti.job_id to None
This helps to avoid clashing with the _find_zombies method which is now made to only act on jobs that are not
running and are also not heartbeating but the task is running.

In a situation where the _find_zombies method and adopt_or_reset_orphaned_task method run at the same time,
there would be conflict but now there won't be conflicts because _find_zombies only act on failed jobs with
running tasks and adopt_or_reset_orphaned_tasks kill jobs

ashb reviewed (changes requested):
I think you are working on two/three overlapping problems here -- 

1. When you exit a scheduler using LocalExecutor, the running tasks (i.e. the LocalTaskJob processes) aren't killed too, and the usually should be.
2. LocalExecutor can't addopt tasks (because it expects them to not be running if the scheduler isn't)

The issue might be that we blindly adopt/reset orphaned tasks even though they might still be running and heartbeating.

So this change isn't right I don't think.

Certainly setting `ti.job_id = None` is wrong -- if the ti.job_id has a value then we want to wait until that has been detected as a zombie before changing that value.
